internal/rbd/snapshot.go

```diff
@@ -176,3 +178,131 @@ func undoSnapshotCloning(
 
 	return err
 }
+
+// NewSnapshotByID creates a new rbdSnapshot from the rbdVolume.
+//
+// Parameters:
+// - name of the new rbd-image backing the snapshot
+// - id of the rbd-snapshot to clone
+//
+// FIXME: When resolving the Snapshot, the RbdImageName will be set to the name
+// of the parent image. This is can cause issues when not accounting for that
+// and the Snapshot is deleted; instead of deleting the snapshot image, the
+// parent image is removed.
+func (rv *rbdVolume) NewSnapshotByID(
+	ctx context.Context,
+	cr *util.Credentials,
+	name string,
+	id uint64,
+) (types.Snapshot, error) {
+	snap := rv.toSnapshot()
+	snap.RequestName = name
+
+	// reserveSnap sets snap.{RbdSnapName,ReservedID,VolID}
+	err := reserveSnap(ctx, snap, rv, cr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a reservation in the journal for snapshot image %q: %w", snap, err)
+	}
+	defer func() {
```

Comment:
defer should be defined before the error check, So that, `undoSnapReservation` is called

Reply:
No, that should not be needed. If `reserveSnap()` failed, there is also no need to undo the reservation?

Reply:
`reserveSnap()` can fail  after `ReserveName()`

https://github.com/ceph/ceph-csi/blob/d5849a4801fa9f3383cefc4f96e18fcad420dc80/internal/rbd/rbd_journal.go#L403-L421

Reply:
I would say that is a bug in `reserveSnap()`, there should not be anything to undo by the caller in case of an error. Callers of a function should not need to be aware of the internals of functions (or functions that are called from within the function).

Reply:
Opened #4945 for improving `reserveSnap()`.